src/openrct2/world/banner.cpp

```diff
@@ -0,0 +1,621 @@
+{
+    gCommandExpenditureType = RCT_EXPENDITURE_TYPE_LANDSCAPING;
+    sint32 z = (baseHeight * 8);
+    gCommandPosition.x = x + 16;
+    gCommandPosition.y = y + 16;
+    gCommandPosition.z = z;
+
+    if (!(gScreenFlags & SCREEN_FLAGS_SCENARIO_EDITOR) && !gCheatsSandboxMode)
+    {
+        if (!map_is_location_owned(x, y, z - 16))
+        {
+            return MONEY32_UNDEFINED;
+        }
+    }
+
+    if (flags & GAME_COMMAND_FLAG_APPLY)
+    {
+        rct_map_element* mapElement = map_get_first_element_at(x / 32, y / 32);
+
+        bool found = false;
+        do
+        {
+            if (map_element_get_type(mapElement) != MAP_ELEMENT_TYPE_BANNER)
+                continue;
+
+            if (mapElement->properties.banner.position != direction)
+                continue;
+
+            found = true;
+            break;
+        } while (!map_element_is_last_for_tile(mapElement++));
+
+        if (found == false)
+        {
+            return MONEY32_UNDEFINED;
+        }
+
+        rct_window* window = window_find_by_number(WC_BANNER, mapElement->properties.banner.index);
+        if (window)
+        {
+            window_invalidate(window);
+        }
+        gBanners[mapElement->properties.banner.index].colour = colour;
+        map_invalidate_tile_zoom1(x, y, z, z + 32);
+    }
+
+    return 0;
+}
+
+static money32 BannerPlace(sint16 x, sint16 y, uint8 pathBaseHeight, uint8 direction, uint8 colour, uint8 type, uint8 *bannerIndex, uint8 flags)
+{
+    gCommandPosition.x = x + 16;
+    gCommandPosition.y = y + 16;
+    gCommandPosition.z = pathBaseHeight * 16;
+    gCommandExpenditureType = RCT_EXPENDITURE_TYPE_LANDSCAPING;
+    if (game_is_paused() && !gCheatsBuildInPauseMode)
+    {
+        gGameCommandErrorText = STR_CONSTRUCTION_NOT_POSSIBLE_WHILE_GAME_IS_PAUSED;
+        return MONEY32_UNDEFINED;
+    }
+
+    if (!map_check_free_elements_and_reorganise(1))
+    {
+        return MONEY32_UNDEFINED;
+    }
+
+    if (x >= 256 * 32 || y >= 256 * 32)
+    {
+        return MONEY32_UNDEFINED;
+    }
+
+    rct_map_element* mapElement = map_get_first_element_at(x / 32, y / 32);
+
+    bool pathFound = false;
+    do
+    {
+        if (map_element_get_type(mapElement) != MAP_ELEMENT_TYPE_PATH)
+            continue;
+
+        if (mapElement->base_height != pathBaseHeight * 2 && mapElement->base_height != (pathBaseHeight - 1) * 2)
+            continue;
+
+        if (!(mapElement->properties.path.edges & (1 << direction)))
+            continue;
+
+        pathFound = true;
+        break;
+    } while (!map_element_is_last_for_tile(mapElement++));
+
+    if (pathFound == false)
+    {
+        gGameCommandErrorText = STR_CAN_ONLY_BE_BUILT_ACROSS_PATHS;
+        return MONEY32_UNDEFINED;
+    }
+
+    if (!(gScreenFlags & SCREEN_FLAGS_SCENARIO_EDITOR) &&
+        !gCheatsSandboxMode &&
+        !map_is_location_owned(x, y, pathBaseHeight * 16))
```

Comment:
This is a very common check, might be good to write a helper now... `map_can_build_at(x, y, z)`